### PR TITLE
New: CPTextField now supports the CPEditorRegistrationProtocol

### PR DESCRIPTION
**Pull request by daboe01**
previously, changing the selection of a CPArrayController whilst editing a bound text field could result in losing these edits.
this PR makes the textfield invoke the objectDidBeginEditing and objectDidEndEditing methods of the CPArrayController appropriately.
additionally, all dirty changes are committed before the selection changes.

fixes issue #1435

### CODE REVIEW

File: AppKit/CPTextField.j [hunk 1]
```diff
@@ -1933,6 +1933,56 @@ var CPTextFieldIsEditableKey            = "CPTextFieldIsEditableKey",
...
+        notificationCenter = [CPNotificationCenter defaultCenter];
+
+    if (theBinding)
+    {
+        [notificationCenter removeObserver:[theBinding._info objectForKey:CPObservedObjectKey]
+                                      name:@"CPControlTextDidBeginEditingNotification"
+                                    object:anObject];
+
+        [notificationCenter removeObserver:[theBinding._info objectForKey:CPObservedObjectKey]
+                                      name:@"CPControlTextDidEndEditingNotification"
```
Comment (Dogild) [inline AppKit/CPTextField.j]: Here you should use the var `name:CPControlTextDidBeginEditingNotification`

File: AppKit/CPTextField.j [hunk 2]
```diff
@@ -1933,6 +1933,56 @@ var CPTextFieldIsEditableKey            = "CPTextFieldIsEditableKey",
...
 
++ (void)unbind:(CPString)aBinding forObject:(id)anObject
+{
+    var theBinding = [self getBinding:aBinding forObject:anObject],
+        notificationCenter = [CPNotificationCenter defaultCenter];
+
+    if (theBinding)
+    {
+        [notificationCenter removeObserver:[theBinding._info objectForKey:CPObservedObjectKey]
+                                      name:@"CPControlTextDidBeginEditingNotification"
```
Comment (Dogild) [inline AppKit/CPTextField.j]: Here you should use the var `CPControlTextDidBeginEditingNotification`

**Comment by Dogild** [comment]
@cacaodev and @daboe01 what's going on with this one ?

**Comment by daboe01** [comment]
imho, we can merge this in.

**Comment by cacaodev** [comment]
This PR effectively solves the issue mentioned but it could be generalized
- The CPEditorRegistrationProtocol can be implemented in all controls that support editing, not just CPTextField.
- In CPArrayController there are other cases than selection change when we need to review all editor pending changes. They should be covered, including the selection change, by the wider concept described by the methods commitEditing: (forces to end editing) and discardEditing: (pending changes are lost).

@daboe01 Do you want to work on this or do you prefer to merge as it is ? In this case, I will just ask you to document the fact that this is a partial implementation. In code comments or by opening a new PR.

**Comment by cacaodev** [comment]
Also, it seems that with this PR, when the array controller selection changes, editor changes are lost. Is it true and did you check the cocoa behavior ?

**Comment by cacaodev** [comment]
Take a look at CPController, there is a commitEditing: method implemented there that calls commitEditing: on the editors. The solution could be as simple as calling it from the _selectionWillChange method in CPArrayController add a commitEditing protocol to CPObject that will just reverseSet the binding.

One last thing: when an editor commits its value, it needs to be removed from the list of active editors in CPController, don't you think ?

**Comment by daboe01** [comment]
@cacaodev thank you for the valuable insight. i will look into all this as soon as possible.
+#needs-improvement

**Comment by daboe01** [comment]
-#needs-review

**Comment by daboe01** [comment]
-#needs-confirmation

**Comment by daboe01** [comment]
@cacaodev it is removed by the call to objectDidEndEditing that is triggered by the CPControlTextDidEndEditingNotification

**Comment by daboe01** [comment]
@cacaodev @mrcarlberg i think it is time to merge this in as it improves the structure and cocoa compatibility of our code-base right now.
i will work on the generalization with future pull requests as soon as the need arises on my side.

**Comment by mrcarlberg** [comment]
@cacaodev What do you think about this PR? How should we move this forward?

**Comment by cacaodev** [comment]
@mrcarlberg As I said before, as far as I know, CPEditorRegistrationProtocol is not really for user consumption but should be used inside CPArrayController.
We can merge this but without a change in CPArrayController I am afraid this will just be a nice sleeping feature (_la belle endormie_ in french).

**Comment by mrcarlberg** [comment]
@cacaodev I do agree that it should be good if this PR can cover more cases and be more generic. But it does give us some basic functions that can be used and built upon for future enhancements. I think we should merge this if we document in the code what is missing.

@daboe01 Can you add comments about the issues @cacaodev talks about above. It is good if we can have this documented in the code.

**Comment by cacaodev** [comment]
@mrcarlberg that's good for me

**Comment by daboe01** [comment]
@cacaodev @mrcarlberg i am a bit unsure where to place the comments and what to comment exactly.

should i add FIXME-comments to all our editable controls that support binding to call objectDidBeginEditing: and objectDidEndEditing?

**Comment by mrcarlberg** [comment]
Yes, you can add FIXME comments on the controls. I also think it is a good idea to add the two issues in the comment from 14 Dec 2016 to some file. Maybe in the CPObjectController.j file above the added method is a good place?

**Comment by daboe01** [comment]
milestone=1.1

**Comment by daboe01** [comment]
-#needs-improvement
+#ready-to-commit

**Comment by daboe01** [comment]
@mrcarlberg can we merge this one in, eventually?

**Comment by mrcarlberg** [comment]
Merged! Thanks!

**Pull request merged**